File: app/src/main/java/com/braincademy/junior/game/WaveGenerator.kt
package com.braincademy.junior.game

import com.braincademy.junior.model.DifficultyConfig
import com.braincademy.junior.model.SnakeSpellConstants
import com.braincademy.junior.model.SnakeType
import kotlin.random.Random

object WaveGenerator {
    fun getSnakeCountForWave(waveNumber: Int): Int {
        return (3 + waveNumber).coerceAtMost(12)
    }

    fun pickSnakeType(
        waveNumber: Int,
        config: DifficultyConfig,
    ): SnakeType {
        if (waveNumber <= 1) return SnakeType.GREEN
        return config.availableSnakeTypes.random()
    }

    fun getSpawnInterval(
        waveNumber: Int,
        config: DifficultyConfig,
    ): Float {
        val reduction = (waveNumber - 1) * 0.1f
        return (config.baseSpawnInterval - reduction).coerceAtLeast(1.0f)
    }

    fun getSnakeSpeed(
        snakeType: SnakeType,
        waveNumber: Int,
        config: DifficultyConfig,
    ): Float {
        val waveBonus = (waveNumber - 1) * 2f
        return (config.baseSnakeSpeed + waveBonus) * snakeType.speedMultiplier
    }

    fun pickLane(): Int = Random.nextInt(SnakeSpellConstants.NUM_LANES)
}

File: UnityProject/BrainAcademy/Assets/Scripts/SnakeSpell/WaveGenerator.cs
using UnityEngine;

public static class WaveGenerator
{
    public static int GetSnakeCountForWave(int waveNumber)
    {
        return Mathf.Min(3 + waveNumber, 12);
    }

    public static SnakeType PickSnakeType(int waveNumber, DifficultyConfig config)
    {
        if (waveNumber <= 1) return SnakeType.Green;
        return config.availableSnakeTypes[Random.Range(0, config.availableSnakeTypes.Count)];
    }

    public static float GetSpawnInterval(int waveNumber, DifficultyConfig config)
    {
        float reduction = (waveNumber - 1) * 0.1f;
        return Mathf.Max(config.baseSpawnInterval - reduction, 1.0f);
    }

    public static float GetSnakeSpeed(SnakeType snakeType, int waveNumber, DifficultyConfig config)
    {
        float waveBonus = (waveNumber - 1) * 2f;
        return (config.baseSnakeSpeed + waveBonus) * SnakeTypeData.GetSpeedMultiplier(snakeType);
    }

    public static float PickAngle()
    {
        return Random.Range(0f, 360f);
    }
}
